shell/common/shell_unittests.cc

```diff
@@ -4570,7 +4574,11 @@ TEST_F(ShellTest, ShellFlushesPlatformStatesByMain) {
     // The construtor for ViewportMetrics{_, width, _, _, _} (only the 2nd
     // argument matters in this test).
     platform_view->SetViewportMetrics(0, ViewportMetrics{1, 10, 1, 0, 0});
-    shell->AddView(1, ViewportMetrics{1, 30, 1, 0, 0});
+    shell->AddView(1, ViewportMetrics{1, 30, 1, 0, 0}, [](bool added) {
+      // The Dart isolate hasn't been started yet. The view won't be added until
+      // the shell flushes the platform state.
+      ASSERT_FALSE(added);
```

Comment:
This is kind of weird. This will make flushing not very usable. Imagine if we add some verifying mechanism to `FlutterViewController` that if the respective view addition fails, the view controller is deemed invalid. With your design, `FlutterViewController` constructed before the Dart isolate is bound to be invalid.

A possible way is to delay the callback until Dart isolate is ready. What do you think?

Reply:
I was considering making `RuntimeController::AddView` return `true` if:

1. The isolate is launched and the view was added, or...
2. The isolate isn't launched yet, and the view was queued

When the isolate is launched, `RuntimeController::FlushRuntimeStateToIsolate` returns `false` if the queued views couldn't be created (this is existing behavior).

Does that seem reasonable?

Reply:
> This will make flushing not very usable.

Could you expand on why you say this?

> With your design, `FlutterViewController` constructed before the Dart isolate is bound to be invalid.

Right. This follows the multi-view embedder API design: https://flutter.dev/go/multi-view-embedder-apis#heading=h.v5bj81fmonnl

> If `FlutterEngineAddView` is called when the engine is not running, the engine will return `kInternalInconsistency`.

You must launch the engine before you can call `FlutterEngineAddView` and `FlutterEngineRemoveView`.

Reply:
I'm not saying before the engine is run. After the engine is run, it takes a while before the Dart isolate is ready (basically when the Dart `main` function starts). The app should be able to add views during this gap.

Take the following snippet that we're currently using in the multiview MVP for example:
```swift
  override func applicationDidFinishLaunching(_ aNotification: Notification) {
    // Engine is run here, but Dart isolate is not ready
    flutterEngine.run(withEntrypoint: nil)
    RegisterGeneratedPlugins(registry: self.flutterEngine)

    let mainWindow = self.mainFlutterWindow as! MainFlutterWindow
    mainWindow.showFlutter(engine: flutterEngine)

    // A second window is created, which adds a second view controller 
    // before the Dart isolate is ready.
    let newWindow = SideFlutterWindow(engine: self.flutterEngine)
  }
```
Since we aim at a synchronous runner API, it'll be more intuitive and convenient to allow the user to add view controllers immediately after creating the engine.

Reply:
> but after the engine is run and before the Dart isolate is ready. The gap takes quite a while, and the app should be able to add views during this gap.

That's an excellent point, `FlutterEngineRun` does not block until the Dart isolate is launched. Let me think on this some more, but I think I agree with your suggestion above:

> A possible way is to delay the callback until Dart isolate is ready.

Reply:
@dkwingsmt Hey there, I updated this with your suggestion to delay the callback until the Dart isolate is launched. Let me know what you think!

Reply:
@dkwingsmt FYI, I'm moving this to a separate pull request: https://github.com/flutter/engine/pull/51659

Feel free to leave comments there!

Reply:
This was moved to: https://github.com/flutter/engine/pull/51659